syntax = "proto3";
import "proto/GraphDef.proto";
import "proto/KernelDef.proto";
import "proto/GraphVis.proto";

package navi;
option cc_enable_arenas = true;

message EdgeOverrideDataDef
{
    int32 graph_id = 1;
    NodePortDef output_port = 2;
    repeated int32 indexs = 3;
}

message OverrideDataValueDef {
    bytes type = 1;
    bytes data = 2;
}

message OverrideDataDef
{
    repeated OverrideDataValueDef datas = 1;
    repeated EdgeOverrideDataDef edge_overrides = 2;
}

message NamedDataDef
{
    int32 graph_id = 1;
    string name = 2;
    bytes type = 3;
    bytes data = 4;
}

message TraceBtFilter
{
    string file = 1;
    int32 line = 2;
}

message RunParams
{
    SessionIdDef id = 1;
    uint32 thread_limit = 2;
    int64 timeout_ms = 3;
    string trace_level = 4;
    bool collect_metric = 5;
    bool collect_perf = 6;
    int32 max_inline = 7;
    OverrideDataDef override_data = 8;
    repeated TraceBtFilter trace_bt_filters = 9;
    string task_queue_name = 10;
    repeated NamedDataDef named_datas = 11;
    ResourceStage resource_stage = 12;
}

message NaviPortData
{
    int32 port_id = 1;
    int32 to_part_id = 2;
    int32 from_part_id = 3;
    bytes type = 4;
    bool eof = 5;
    bytes data = 6;
}

message BorderIdDef {
    int32 graph_id = 1;
    int32 peer = 2;
    int32 io_type = 3;
}

message NaviBorderData
{
    BorderIdDef border_id = 1;
    repeated NaviPortData port_datas = 2;
}

enum NaviCompressType {
    CT_NONE = 0;
    CT_LZ4 = 1;
}

message DomainGraphDef {
    bytes sub_graph = 1;
    int32 this_part_id = 2;
    NaviCompressType compress_type= 3;
    GraphCounterInfo counter_info = 4;
}

message NaviMessage
{
    uint32 msg_id = 1;
    DomainGraphDef domain_graph = 2;
    RunParams params = 3;
    repeated NaviBorderData border_datas = 4;
    SerializedResultDef navi_result = 5;
    repeated SerializedResultDef sub_result = 6;
    int64 version = 7;
}
